File: python/CNTtools/tools/bandpower.py
from scipy.signal import welch
from scipy.integrate import simpson
import numpy as np
from beartype import beartype
from beartype.typing import Union, Iterable
from numbers import Number


@beartype
def bandpower(
    data: np.ndarray,
    fs: Number,
    band: Iterable[Number],
    win_size: Union[None, Number] = None,
    relative: bool = False,
) -> np.ndarray:
    """Adapted from https://raphaelvallat.com/bandpower.html
    Compute the average power of the signal x in a specific frequency band.

    Parameters
    ----------
    data : 1d-array or 2d-array
        Input signal in the time-domain. (time by channels)
    fs : float
        Sampling frequency of the data.
    band : list
        Lower and upper frequencies of the band of interest.
    win_size : float
        Length of each window in seconds.
        If None, win_size = (1 / min(band)) * 2
    relative : boolean
        If True, return the relative power (= divided by the total power of the signal).
        If False (default), return the absolute power.

    Return
    ------
    bp : np.ndarray
        Absolute or relative band power. channels by bands
    """
    band = np.asarray(band)
    assert len(band) == 2, "CNTtools:invalidBandRange"
    assert band[0] < band[1], "CNTtools:invalidBandRange"
    nchan = data.shape[1]
    bp = np.nan * np.zeros(nchan)
    low, high = band

    # Define window length
    if win_size is not None:
        nperseg = int(win_size * fs)
    else:
        nperseg = int((2 / low) * fs)

    # Compute the modified periodogram (Welch)
    freqs, psd = welch(data.T, fs, nperseg=nperseg)

    # Frequency resolution
    freq_res = freqs[1] - freqs[0]

    # Find closest indices of band in frequency vector
    idx_band = np.logical_and(freqs >= low, freqs <= high)

    # Integral approximation of the spectrum using Simpson's rule.
    if psd.ndim == 2:
        bp = simpson(psd[:, idx_band], dx=freq_res)
    elif psd.ndim == 1:
        bp = simpson(psd[idx_band], dx=freq_res)

    if relative:
        bp /= simpson(psd, dx=freq_res)

    return bp

File: matlab/tools/calc_features/band_power.m
function bp = band_power(data, fs, band, varargin)
% Adapted from https://raphaelvallat.com/bandpower.html
% Compute the average power of the signal x in a specific frequency band.
%
%     Parameters
%     ----------
%     data : matrix
%         Input signal in the time-domain. (time by channels)
%     fs : float
%         Sampling frequency of the data.
%     band : list
%         Lower and upper frequencies of the band of interest. 
%     win : float
%         Length of each window in seconds.
%         If None, win = (1 / min(band)) * 2
%     relative : boolean
%         If True, return the relative power (= divided by the total power of the signal).
%         If False (default), return the absolute power.
%
%     Return
%     ------
%     bp : matrix, channels X bands
%         Absolute or relative band power.

defaults = {[],false};

for i = 1:length(varargin)
    if isempty(varargin{i})
        varargin{i} = defaults{i};
    end
end

p = inputParser;
addRequired(p, 'data', @isnumeric);
addRequired(p, 'fs',  @isnumeric);
addRequired(p, 'band', @isnumeric);
addOptional(p, 'win', defaults{1}, @isnumeric);
addOptional(p, 'relative', defaults{2}, @islogical);
parse(p, data, fs, band, varargin{:});

% Access parsed values
data = p.Results.data;
fs = p.Results.fs;
band = p.Results.band;
win = p.Results.win;
relative = p.Results.relative;

if ~isempty(win)
    nperseg = round(win * fs);
else
    nperseg = [];%round((2 / low) * fs);
end
% Compute the modified periodogram (Welch)
[pxx, f] = pwelch(data, nperseg, [], [], fs);
% Frequency resolution
% freq_res = f(2) - f(1);

% Find closest indices of band in frequency vector
% idx_band = find(f >= low & f <= high);

bp = bandpower(pxx,f,band,'psd');

if relative
    bp = bp./bandpower(pxx,f,'psd');
end
